akka-projection-slick/src/main/scala/akka/projection/slick/internal/SlickProjectionImpl.scala

```diff
@@ -0,0 +1,69 @@
+/*
+ * Copyright (C) 2020 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.projection.slick.internal
+
+import akka.Done
+import akka.actor.ClassicActorSystemProvider
+import akka.annotation.InternalApi
+import akka.event.Logging
+import akka.projection.Projection
+import akka.projection.slick.OffsetStore
+import akka.stream.scaladsl.{ Keep, Sink, Source }
+import akka.stream.{ KillSwitch, KillSwitches }
+import slick.basic.DatabaseConfig
+import slick.dbio.DBIO
+import slick.jdbc.JdbcProfile
+
+import scala.concurrent.Future
+@InternalApi
+private[projection] class SlickProjectionImpl[Offset, StreamElement, P <: JdbcProfile](
+    projectionId: String,
+    sourceProvider: Option[Offset] => Source[StreamElement, _],
+    offsetExtractor: StreamElement => Offset,
+    databaseConfig: DatabaseConfig[P],
+    eventHandler: StreamElement => DBIO[Done])
+    extends Projection {
+
+  private val offsetStore = new OffsetStore(databaseConfig.db, databaseConfig.profile)
+
+  private var shutdown: Option[KillSwitch] = None
+
+  override def start()(implicit systemProvider: ClassicActorSystemProvider): Unit = {
+
+    // TODO: add a LogSource for projection when we have a name and key
+    val akkaLogger = Logging(systemProvider.classicSystem, this.getClass)
+
+    implicit val dispatcher = systemProvider.classicSystem.dispatcher
+    import databaseConfig.profile.api._
+
+    val lastKnownOffset: Future[Option[Offset]] = offsetStore.readOffset(projectionId)
+
+    val killSwitch =
+      Source
+        .future(lastKnownOffset.map { offsetOpt =>
+          akkaLogger.debug("Starting projection '{}' from offset '{}'", projectionId, offsetOpt)
+          sourceProvider(offsetOpt)
+        })
+        .flatMapConcat[StreamElement, Any](identity)
+        .viaMat(KillSwitches.single)(Keep.right)
+        .mapAsync(1) { elt =>
+          // run user function and offset storage on the same transaction
+          // any side-effect in user function is at-least-once
+          val txDBIO =
+            offsetStore.saveOffset(projectionId, offsetExtractor(elt)).flatMap(_ => eventHandler(elt))
+
+          databaseConfig.db.run(txDBIO.transactionally)
```

Comment:
It seems that we swallow the `DBIO` returned by the user flow for each offset.  I would expect it to fail the inner stream and made available to the user somehow.

Reply:
Yep, that's the issue we found on the testkit. It's fixed now.

Reply:
I think this case is a little different. I looked at `SlickProjectionSpec` test `"restart from previous offset - fail with DBIOAction.failed"`, but I don't see how the failure case results in a `DBIOAction.failed`. The  exception is raised within the assert function itself.

If a `DBIOAction.failed` occurs as a result of an error returned by the DB driver then I _think_ it's translated into a `Failure`, but it's not clear.

Reply:
This have already changed. Now, the testkit is giving preference to the any exception coming from the stream. 

The current code is as following:

```scala
 val streamFailure =
  intercept[RuntimeException] {
    projectionTestKit.run(slickProjectionFailing) {
      fail("fail assertFunc, we want to capture the stream failure")
    }
  }
```
(https://github.com/akka/akka-projection/pull/35/files#diff-244b95b819c7d32f7f7ff7881683bf9eR116-R121)

The `assertFunc` is even not trying to do anything useful. 

The projection fails because we use `DBIOAction.failed` and indeed, that will cause the `Future` returned by Slick to be a `Failure`. 

Then, later in the test, we check that `streamFailure` contains the message we used in the `DBIOAction.failed`. 

https://github.com/akka/akka-projection/pull/35/files#diff-244b95b819c7d32f7f7ff7881683bf9eR123-R125

Reply:
Got it. Thanks for the clarification.